Allow importing of memory.

koponen-styra opened the pull request:
This is still missing unit tests but wanted to get feedback first whether this is the right direction to add support for memory imports.

Hywan commented:
For what I read, the patch is in good tracks. We want to abandon the old import API and focus on the `ImportObject` API. That's the correct way to go.

Thanks for working on it!

koponen-styra commented:
Do you mean instead of adding AppendFunction to Imports, have another constructor for ImportObject to construct a memory import? Sorry, just lacking still a lot of context here. Concrete suggestions are much appreciated.

Hywan commented:
The idea is to use https://wasmerio.github.io/wasmer/c/runtime-c-api/wasmer_8h.html#a9b254ed5edbb735de3e9c2a83ddf6040 by passing https://wasmerio.github.io/wasmer/c/runtime-c-api/structwasmer__import__t.html with `tag` set to `WASM_MEMORY`. And so you need to update https://github.com/wasmerio/go-ext-wasm/blob/master/wasmer/import.go#L67 to support any kind of imports, not just `WASM_FUNCTION`.

koponen-styra commented:
The PR actually already changes the ImportObject's Extend to handle wasmer_import_t with memory tag, by improving getCWasmerImport to handle the memory imports as well. So perhaps after all, the main thing missing are tests.

Hywan commented:
I'll review your PR as soon as possible.

Hywan commented on wasmer/test/imports.go:
```diff
@@ -292,3 +292,38 @@ func testWasiImportObject(t *testing.T) {
+	err = importObject.Extend(*imports)
+	assert.NoError(t, err)
+
+	instance, err := module.InstantiateWithImportObject(importObject)
```
I'm surprised it works. See my review comment. The test doesn't actually test the memory in this case.

Hywan commented on wasmer/test/testdata/log_import_memory.wat:
```diff
@@ -0,0 +1,15 @@
+  (import "env" "log_message" (func (;0;) (type 0)))
+  (import "env" "memory" (memory (;0;) 17))
+  (func (;1;) (type 1)
+    i32.const 1048576
```
We don't read from the memory here, but from the statically allocated data, i.e. from the Data Section.

koponen-styra replied:
Ouch. Interesting it works... I'll fix it to read from memory.

Hywan commented:
I've opened https://github.com/koponen-styra/go-ext-wasm/pull/1 after my reviews. That's a small set of improvements.

Hywan commented:
To be clear: Apart from the test that doesn't test the new feature, the code looks really great, good job!

koponen-styra commented on wasmer/memory.go:
```diff
@@ -86,3 +108,8 @@ func (memory *Memory) Grow(numberOfPages uint32) error {
+
+// Close closes/frees memory allocated at the NewMemory at time.
+func (memory *Memory) Close() {
+	cWasmerMemoryDestroy(memory.memory)
```
Makes sense.

koponen-styra replied:
Resolving this since it's a part of the other PR.

koponen-styra commented:
Let me know if you prefer me to pull in your commits from your PR on top of this one or whether you prefer the other way around.

koponen-styra commented:
Fixed the test and made it minimal. In addition, noticed and fixed a minor issue with importObject.Imports: it considered modules with no function imports as invalid.

syrusakbary commented:
@koponen-styra if you can pull @Hywan commits into this PR that would be great! It will accelerate a bit the process :)

Hywan commented on wasmer/import.go:
```diff
@@ -35,11 +35,6 @@ func NewImportObject() *ImportObject {
 func (importObject *ImportObject) Imports() (*Imports, error) {
 	imports := cWasmerImportObjectGetFunctions(importObject.inner)
-
-	if imports == nil {
```
This is dangerous, but we need to fix this anyway.

Hywan commented:
bors r+
